source/transfer_list.rst

```diff
@@ -902,5 +902,42 @@ This manifest contains the SPMC attribute node consumed by the SPMD
      - hdr_size
      - Holds a SPMC manifest image in DT format.
 
+
+**Executable entry point information entry layout (XFERLIST_EXEC_EP_INFO)**
+
+This entry type holds an array of `entry_point_info_t` structures.
+`entry_point_info_t` is a data structure containing information used by a
+bootloader to initialize the execution state of an image. This entry provides
+the consumer with all information to execute subsequent images.
+
+.. _tab_entry_point_info:
+.. list-table:: Entry point info type layout
+   :widths: 2 2 2 8
+
+   * - Field
+     - Size (bytes)
+     - Offset (bytes)
+     - Description
+
+   * - tag_id
+     - 0x3
+     - 0x0
+     - The tag_id field must be set to **0x102**.
+
+   * - hdr_size
+     - 0x1
+     - 0x3
+     - |hdr_size_desc|
+
+   * - data_size
+     - 0x4
+     - 0x4
+     - The size of SPMC manifest in bytes.
+
+   * - ep_info
+     - data_size
```

Comment:
the DT takes up the whole entry, so we should be able to drop data_size

Reply:
Not sure what you mean. data_size is a standard TE header field. The ep_info array is of variable size and even if it wasn't, data_size enables us to add new fields later.

Reply:
Oh right, but then why is the size variable? Shouldn't it always be 4? If this is extended in future, then we would want a standard size for spmc_man

Reply:
I'm a bit confused by the mention of spmc_man when this is about ep_info. The size of a single ep_info can be discovered but this is an array of variable size and data_size enables us to determine how many entries are in the array. However, by using data_size in this way, it prevents us adding new fields to this TE in future. So I'm now thinking it might be better to add an "ep_info array size" field to enable future extension.

Reply:
> I'm a bit confused by the mention of spmc_man when this is about ep_info.

I think the confusion stems from an earlier diff that the original comment was applied to, this referred to spmc_man rather than ep_info.

> However, by using data_size in this way, it prevents us adding new fields to this TE in future. So I'm now thinking it might be better to add an "ep_info array size" field to enable future extension.

To address the issue of versioning, could a consumer of the TE not use type punning to interpret the the first header in the blob? By doing this, they have access to all the information they need to handle different versions of the data structure, including its size. Admittedly, this might introduce some additional complexity on the consumer side.

Reply:
Yes, clients can inspect the ep_info header to determine its size (for a given version), but that won't tell you how many elements are in the array. This can be inferred from data_size if there are no other fields but if we add new fields later, an old consumer might try to interpret those new fields as another ep_info. I suppose it might be OK if the sizeof (new fields)  < sizeof(ep_info) but it's a bit flaky to rely on that.

Reply:
> but if we add new fields later, an old consumer might try to interpret those new fields as another ep_info.

That's a good point. If I’m not mistaken, you’re saying that we may at some point want to populate the data field with some additional information similar to the flags field `XFERLIST_EVLOG`? In which case, the data_size will not be a proxy variable for the ep_info count.

Reply:
Instead of count, how about having a separate TE for each piece of info? The contents should have enough info to know which state it relates to, right?

Reply:
If everyone thinks writing parsers/APIs that can deal with multiple TEs which have the same tag is too complex, we should maybe change the spec to disallow them after all. It wouldn't be good if some parts of the ecosystem can support and use those constructs and others always try to find an alternative solution because their TL code can't handle it. I originally wrote the section that says it's allowed, but I admit I can't really come up with many practical examples where an array of structs within a single TE wouldn't do just as well. There may be the occasional edge case where a TE needs to be moved and expanded to add more entries to the array, but it should be pretty rare and maybe the ease of use of supporting a "get entry by tag ID" API is worth that.

Reply:
After discussing with Manish and experimenting with our library implementation, we’ve decided to adopt the single EP per TE approach proposed by @sjg20. It appears that managing multiple TEs with the same tag ID in the transfer list is simpler than initially anticipated. Overall, packing the structures into an array introduces complexity for both the sender and receiver, especially if we need to include additional metadata solely to convey information about the array.

Reply:
I'm glad you've managed to make the single EP for TE work.

Now there is only a single ep_info in this field, I tend to agree with Simon that saying the size of this field is data_size implies that this is of variable size when in fact it will always be sizeof(ep_info). So perhaps better to say that instead since it will become incorrect if this TE is ever extended.

I notice the same issue exists in the pp_addr and spmc_man fields of previous TEs. It would be nice If you can fix those as well (in a separate commit) but not critical for this PR.

Reply:
Thank you, @danh-arm, for the feedback. I’ve created a separate pull request to address the same issue for pp_addr. Since spmc_man is a device tree, using data_size might be more suitable in this context.

https://github.com/FirmwareHandoff/firmware_handoff/pull/35

Reply:
@raymo200915 recently pointed me to this thread. I think there are very valid concerns raised for having multiple entries with the same ID. AFAICT you are covering a very specific use case here, were everything is consumed by TF-A. But in the future people might want to add multiple DT entries for example -- Imagine a single bootloader that supports multiple SoCs and the only difference is the DT.
In that case blindly adding multiple entries is not enough. We need to be able to uniquely identify an entry.

I'll go through the comments again, but I think this needs some fixing

Reply:
@apalos 
By looking closer to this issue, I think the user has two options when they need multiple TEs for same or similar purpose.

If the caller of each boot stage has the knowledge on which entry it should take (e.g. the current scenario of TF-A), then,
Optioin 1: Create multiple TEs with same tag ID, then get the array of entries by inputing the tag ID.

Otherwise,
Option 2: Define new entries as non-standard ones, so that it has a custom tag ID in the range reserved for non-standard TEs - Adding non-standard TE doesn't affect the spec changes.
This option can be used for the scenario of multiple DTs.